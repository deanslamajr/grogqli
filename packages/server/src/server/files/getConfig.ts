import path from 'path';
import findUp from 'find-up';
import editJsonFile from 'edit-json-file';

const CONFIG_FILE_NAME = 'grogqli.json';

export interface GrogqliConfig {
  recordingsSaveDirectory: string;
  port: number;
  shouldDogFood: boolean;
}
const DEFAULT_CONFIG: GrogqliConfig = {
  recordingsSaveDirectory: path.join(process.cwd(), 'grogqli'),
  port: 4000,
  shouldDogFood: false,
};

type GetGrogqliConfigValue = (configKey: keyof GrogqliConfig) => string;
let getGrogqliConfigValue: GetGrogqliConfigValue | undefined;

export const getConfig = async (): Promise<GetGrogqliConfigValue> => {
  // return early if the response is cached
  if (getGrogqliConfigValue) {
    return getGrogqliConfigValue;
  }

  // recursively try to open a file with filename === CONFIG_FILE_NAME
  // if file doesn't exist, try again in parent directory
  // if file isn't found at root directory, return DEFAULT_CONFIG
  const configLocation = await findUp(
    async (directory) => {
      const hasConfig = await findUp.exists(
        path.join(directory, CONFIG_FILE_NAME)
      );
      return hasConfig ? directory : undefined;
    },
    { type: 'directory' }
  );

  let configFromFile: editJsonFile.JsonEditor;

  if (configLocation) {
    const configFileAbsolutePath = path.join(configLocation, CONFIG_FILE_NAME);
    configFromFile = editJsonFile(configFileAbsolutePath);
  }

  getGrogqliConfigValue = (configKey) => {
    if (configKey === undefined) {
      throw new Error(
        'Invalid use of getGrogqliConfigValue: must provide a key'
      );
    }
    let configValue: any;

    // First, try to resolve the value from the given config
    if (configLocation && configFromFile) {
      configValue = configFromFile.get(configKey);

      // Make 'recordingsSaveDirectory' an abs path
      if (
        configKey === 'recordingsSaveDirectory' &&
        configValue !== undefined
      ) {
        configValue = path.join(configLocation, configValue);
      }
    }

    // If config doesn't have a value, use the default
    if (configValue === undefined) {
      configValue = DEFAULT_CONFIG[configKey];
    }

    return configValue;
  };

  return getGrogqliConfigValue;
};
