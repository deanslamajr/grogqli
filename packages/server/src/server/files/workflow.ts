import shortid from 'shortid';
import editJsonFile from 'edit-json-file';

import { DistributiveOmit } from '../types';
import {
  doesFileExist,
  getWorkflowRecordingFilePath,
  getWorkflowNameMappingFilePath,
  mapObjectToJsonFile,
  WORKFLOWS_NAME_TO_ID_MAPPING_VERSION,
} from './';

// pattern to support versioning this file structure
export type WorkflowNameMappingFileContents = WorkflowNameMappingFileContentsVersion1;
interface WorkflowNameMappingFileContentsVersion1 {
  version: 1;
  workflows: {
    [workflowName: string]: {
      name: string;
      id: string;
    };
  };
}

type AddNewEntryToWorkflowMappingFile = (
  workflowName: string
) => Promise<string>;
export const addNewEntryToWorkflowMappingFile: AddNewEntryToWorkflowMappingFile = async (
  workflowName
) => {
  const pathToWorkflowNameToIdMappingFile = await getWorkflowNameMappingFilePath();
  const workflowNameMappingFile = await editJsonFile(
    pathToWorkflowNameToIdMappingFile
  );

  let newWorkflowId;
  // handle the case where mappings file does not exist
  if (!doesFileExist(workflowNameMappingFile)) {
    const initializedWorkflowNamesMappingFileData: WorkflowNameMappingFileContentsVersion1 = {
      version: WORKFLOWS_NAME_TO_ID_MAPPING_VERSION,
      workflows: {},
    };
    mapObjectToJsonFile(
      initializedWorkflowNamesMappingFileData,
      workflowNameMappingFile
    );
    newWorkflowId = shortid.generate();
  } else {
    let newIdIsNotUnique = true;

    // generate a new workflowId that is unique against the existing set of workflowId's
    do {
      newWorkflowId = shortid.generate();
      const workflowMappings = Object.values<
        WorkflowNameMappingFileContentsVersion1['workflows'][keyof WorkflowNameMappingFileContentsVersion1['workflows']]
      >(workflowNameMappingFile.get('workflows'));
      newIdIsNotUnique = workflowMappings.some(
        // eslint-disable-next-line no-loop-func
        ({ id }) => id === newWorkflowId
      );
    } while (newIdIsNotUnique);
  }

  workflowNameMappingFile.set(`workflows.${workflowName}`, {
    name: workflowName,
    id: newWorkflowId,
  });
  workflowNameMappingFile.save();

  return newWorkflowId;
};

// pattern to support versioning this file structure
// TODO remove version0 after a version2 exists
// this is here to help provide an example of how to do this
// (before there was more than a single version)
interface WorkflowRecordingsFileVersion0 {
  version: 0;
}
export type WorkflowRecordingsFile =
  | WorkflowRecordingsFileVersion0
  | WorkflowRecordingsFileVersion1;
export interface WorkflowRecordingsFileVersion1 {
  id: string;
  version: 1;
  description: string;
  operationRecordings: {
    [opId: string]: {
      opId: string;
      opRecordingId: string;
    };
  };
}

export type WorkflowFileBeforeCreation = DistributiveOmit<
  WorkflowRecordingsFile,
  'id'
>;

export const getWorkflowById = async (
  workflowId: string
): Promise<WorkflowRecordingsFile | null> => {
  if (!workflowId) {
    return null;
  }

  const pathToWorkflow = await getWorkflowRecordingFilePath(workflowId);
  let workflow: WorkflowRecordingsFile;
  try {
    workflow = require(pathToWorkflow);
  } catch (error) {
    return null;
  }

  return workflow;
};
