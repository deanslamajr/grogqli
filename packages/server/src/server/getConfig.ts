import path from 'path';
import findUp from 'find-up';

const CONFIG_FILE_NAME = 'grogqli.json';

interface GrogqliConfig {
  recordingsSaveDirectory: string;
  recordingsFilename: string;
}

const DEFAULT_CONFIG: GrogqliConfig = {
  recordingsSaveDirectory: '/home/dhan/codes/grogqli',
  recordingsFilename: 'groql_rec'
};
export const getConfig = async (): Promise<GrogqliConfig> => {
  console.log('process.cwd()', process.cwd());
  // TODO
  // recursively try to open a file with filename === CONFIG_FILE_NAME
  // if file doesn't exist, try again in parent directory
  // if file isn't found at root directory, return DEFAULT_CONFIG
  const configLocation = await findUp(CONFIG_FILE_NAME);

  console.log('configLocation', configLocation)

  // TODO
  // if a config doesn't have a value defined for a particular lookup
  // use the value found in DEFAULT_CONFIG
  return DEFAULT_CONFIG;
}