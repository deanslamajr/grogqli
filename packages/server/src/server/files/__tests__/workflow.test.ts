import path from 'path';

import { getWorkflowById } from '../workflow';
import { getConfig } from '../getConfig';

jest.mock('../getConfig');

describe('workflow', () => {
  beforeEach(() => {
    const mockedGetConfig = getConfig as jest.MockedFunction<typeof getConfig>;
    mockedGetConfig.mockImplementation(async () => () => {
      return path.join(__dirname, 'grogqli');
    });
  });

  describe('getWorkflowById', () => {
    it('should return the given workflow file contents', async () => {
      const actual = await getWorkflowById('someWorkflowId');
      expect(actual).toMatchSnapshot();
    });

    describe('if workflowId is not passed', () => {
      it('should return null', async () => {
        const actual = await getWorkflowById('');
        expect(actual).toBeNull();
      });
    });

    describe('if the file doesnt exist', () => {
      it('should return null', async () => {
        const actual = await getWorkflowById('workFlowDoesntExist');
        expect(actual).toBeNull();
      });
    });
  });
});
