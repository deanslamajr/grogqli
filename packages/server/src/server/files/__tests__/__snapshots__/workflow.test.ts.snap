// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`workflow getWorkflowById should return the given workflow file contents 1`] = `
Object {
  "description": "some markdown text that gives context on this workflow",
  "id": "someWorkflowId",
  "name": "Some workflow",
  "operationRecordings": Object {
    "llK98u0": Object {
      "opRecordingId": "q7f66d54",
    },
    "someOpId": Object {
      "opRecordingId": "someOpRecordingId",
    },
  },
  "version": 1,
}
`;
